File: 프로그래머스/lv2/42584. 주식가격/주식가격.cpp
#include <string>
#include <vector>

using namespace std;

vector<int> solution(vector<int> prices)
{
    vector<int> answer;

    for (int i = 0; i < prices.size(); i++)
    {
        int tmp = prices[i];
        int count = 0;
        if (i == prices.size() - 1)
        {
            answer.push_back(0);
            break;
        }
        for (int j = i + 1; j < prices.size(); j++)
        {
            count++;
            if (tmp > prices[j])
            {
                answer.push_back(count);
                break;
            }
            else if (j == prices.size() - 1)
                answer.push_back(count);
        }
    }

    return answer;
}

File: 프로그래머스/2/42584. 주식가격/주식가격.java
import java.util.*;

class Solution {
    public int[] solution(int[] prices) {
        int[] answer = new int [prices.length];
        Deque<Integer> stk = new ArrayDeque<>();
        
        for(int i = 0; i < prices.length; i++) {
            while (!stk.isEmpty() && prices[stk.peek()] > prices[i]) { 
                int idx = stk.pop(); 
                answer[idx] = i - idx; 
            }
            stk.push(i);
        }
        
        while (!stk.isEmpty()) {
            int idx = stk.pop();
            answer[idx] = prices.length - idx - 1;
        }
        
        return answer;
    }
}
/**

**/